flax/linen/attention.py

```diff
@@ -313,15 +332,15 @@ def __call__(
     if self.decode:
       # detect if we're initializing by absence of existing cache data.
       is_initialized = self.has_variable('cache', 'cached_key')
-      cached_key = self.variable(
-          'cache', 'cached_key', jnp.zeros, key.shape, key.dtype
-      )
-      cached_value = self.variable(
-          'cache', 'cached_value', jnp.zeros, value.shape, value.dtype
-      )
-      cache_index = self.variable(
-          'cache', 'cache_index', lambda: jnp.array(0, dtype=jnp.int32)
-      )
+      cached_key = variable_with_axes('cache', 'cached_key',
```

Comment:
`param_with_axes` and `variable_with_axes` is from an older, deprecated API for sharding (we need to delete it or just move it into the t5x repo which is the primary user).  The newer approach can be found in `core/meta.py` and `linen/spmd.py` and relies on using "boxed" variables that are created by wrapping array initializers.